[TestFix] ExplainLifecycleIT testStepInfoPreservedOnAutoRetry failing

lukewhiting opened the pull request:
Fix for [ES#114220 - [CI] ExplainLifecycleIT testStepInfoPreservedOnAutoRetry failing](https://github.com/elastic/elasticsearch/issues/114220)

After running unstable test for 500+ iterations using random parameters from failing runs, I was unable to reproduce the failure but suspect it's possibly due to a timeout being too short in the busy CI runs.

For now, I have unmuted the test, extended the timeout of the `assertBusy` that checks for the ILM run retries and added additional logging to print out the last response in the event the `assertBusy` times out to assist in future debugging if the extended timeout does not fix the issue.

lukewhiting committed:
Extend timeout of test and add logging on fail

lukewhiting committed:
Unmute unstable test

lukewhiting committed:
Switch to using logger for output

Keeps the forbiddenApis check happy

dakrone commented on x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/ExplainLifecycleIT.java:
```diff
@@ -305,16 +307,23 @@ public void testStepInfoPreservedOnAutoRetry() throws Exception {
+                    containsString("rollover_alias [" + aliasName + "] does not point to index [" + indexName + "]")
+                );
+            }, 30, TimeUnit.SECONDS);
+        } catch (AssertionError e) {
+            logger.error("Assertion failed with response: {}", explainIndexRef.get());
+            throw e;
+        }
```
If you didn't want to catch the AssertionError (we generally try never to catch any Error classes, only Exceptions), it may be better to put a logging statement at `info` level inside of the `assertBusy()`, which would log the output each time (and show up in the logs if it fails again). Then you wouldn't need the `AtomicReference` either.

lukewhiting replied:
That is how I initially implemented it however it really did pollute the logs and looked a bit confusing as it logged out during each attempt at the assertBusy.

I also thought about adding it as a message param to each assertion however those aren't lazy evaluated like a log message param so it was adding to the execution time to do the map.toString for each assertion when it was 99% of the time not actually used.

In the end I decided to optimized for execution time (I'm always conscious time == money in the cloud CI environment) and conciseness but happy to go for logging it out each time if you feel strongly about not catching errors :-)

dakrone replied:
I think it'd be nice to avoid the error catching if possible, so I would advocate for adding the message param to the assertions you think are most likely to fail (`assertBusy`'s back-off gets longer and longer, so I don't think it'd be too bad). Then it's only returned if the entire thing fails.

lukewhiting replied:
Switched to using a message param on the assertions and saving the message to a var so it's only computed once per run

dakrone reviewed (approved):
LGTM, I left one comment

lukewhiting committed:
Switch to using assertion messages to display

To display debug info

lukewhiting committed:
Adjust logic of previous step info preservation

Add additional checks to ensure previous step info can't be cleared
when auto retrying, only updated with new info.

Also added logic to ensure previous step info is cleared when
transitioning to a new action

lukewhiting commented:
Seems the dice roll was in our favour an the IT in question actually failed on the latest CI run, despite the longer timeout so I could take a look at the full state of things: https://gradle-enterprise.elastic.co/s/lituf73fyko2o/console-log/task/:x-pack:plugin:ilm:qa:multi-node:javaRestTest?anchor=57&page=1

lukewhiting committed:
Merge branch 'main' of github.com:elastic/elasticsearch into 114220-preservestepinfo-testfail

lukewhiting commented:
buildkite test this

lukewhiting committed:
Merge branch 'main' of github.com:elastic/elasticsearch into 114220-preservestepinfo-testfail

lukewhiting committed:
Undo accidentally added lines from merge

dakrone commented:
> I have pushed a new commit that adds addition logic to explicitly prevent the previous step info getting nulled out if the state update is a step auto retry or a move to an error state.

I think this is useful, it means that we wouldn't remove the step index until we move forward, which I think more closely matches the original request.

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/114802)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/114802">](https://github.com/elastic/elasticsearch/pull/114802) |
